libs/utils/src/http/endpoint.rs

```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
+                        self.tx.reserve().await.map_err(|_| ())?;
+
+                        // now the response task has picked up the buffer and hopefully started
+                        // sending it to the client.
+                        Ok(())
+                    });
+                    if res.is_err() {
+                        return Err(std::io::ErrorKind::BrokenPipe.into());
+                    }
+                    self.written += n;
+                }
+                Ok(n)
+            }
 
-    let response = Response::builder()
-        .status(200)
-        .header(CONTENT_TYPE, encoder.format_type())
-        .body(Body::from(buffer))
-        .unwrap();
+            fn flushed_bytes(&self) -> usize {
+                self.written
+            }
+        }
 
-    Ok(response)
+        impl std::io::Write for ChannelWriter {
+            fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+                let remaining = self.buffer.capacity() - self.buffer.len();
+
+                let out_of_space = remaining < buf.len();
+
+                if out_of_space {
+                    self.flush0()?;
+                }
+
+                // assume that this will often under normal operation just move the pointer back to the
+                // beginning of allocation, because previous split off parts are already sent and
+                // dropped.
+                self.buffer.extend_from_slice(buf);
+                Ok(buf.len())
+            }
+
+            fn flush(&mut self) -> std::io::Result<()> {
+                self.flush0().map(|_| ())
+            }
+        }
+
+        let mut writer = ChannelWriter::new(128 * 1024, tx);
+
+        let encoder = TextEncoder::new();
+
+        let response = Response::builder()
+            .status(200)
+            .header(CONTENT_TYPE, encoder.format_type())
+            .body(body)
+            .unwrap();
+
+        let span = info_span!("blocking");
+
+        tokio::task::spawn_blocking(move || {
+            let _span = span.entered();
+            // Currently we take a lot of mutexes while collecting metrics, so it's
+            // better to spawn a blocking task to avoid blocking the event loop.
+            let metrics = metrics::gather();
+            let res = encoder
+                .encode(&metrics, &mut writer)
+                .and_then(|_| writer.flush().map_err(|e| e.into()));
+
+            match res {
+                Ok(()) => {
+                    tracing::info!(
+                        bytes = writer.flushed_bytes(),
+                        elapsed_ms = started_at.elapsed().as_millis(),
+                        "responded /metrics"
+                    );
+                }
+                Err(e) => {
+                    tracing::warn!("failed to write out /metrics response: {e:#}");
+                    // semantics of this error are quite... unclear. we want to error the stream out to
```

Comment:
For the non-tokio mpsc, it's documented [more explicitly](https://doc.rust-lang.org/nightly/std/sync/mpsc/struct.SendError.html), but I think for tokio, too, it only gives an error if the receiver has been closed. Not 100% sure tho, haven't dug deep enough into tokio's code. https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#errors

Reply:
Since it's an `mpsc` multiple producer single consumer it will error out. With this confusing comment I was after the semantics of what *hyper* will do upon seeing an Error. I recall it will do nothing, maybe it now closes the connection.

But it matters very little for `/metrics`. We are already saving a lot of allocations with this.